### PR TITLE
Feature/651/configurable temp path

### PR DESCRIPTION
**Pull request by initard**
## Proposed changes

Temp path can now be set via 

```shell
sudo tedge config set tmp.path "/a/directory"
```
Documentation added explaining `tedge-agent` user and group must be added to the folder. 

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by initard** [comment]
I have a question regarding the naming `tmp.path`. Should it be `temp.path` instead?

File: tests/PySys/software_management_end_to_end/sm_apt_install_download_path/pysystest.xml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
+<pysystest type="auto">
+
+  <description>
+	<title>Install package to download path
```
Comment (PradeepKiruvale) [inline tests/PySys/software_management_end_to_end/sm_apt_install_download_path/pysystest.xml:5]: Better to give some more description about the test.

File: tests/PySys/tedge/tedge_software_update/pysystest.xml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
+<pysystest type="auto">
+
+  <description>
+	<title>Software install from download path</title>
```
Comment (PradeepKiruvale) [inline tests/PySys/tedge/tedge_software_update/pysystest.xml:5]: Better to give some more description about the test.

File: tests/PySys/software_management_end_to_end/sm_apt_install_download_path/run.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
+import os
+import subprocess
+from pysys.basetest import BaseTest
+from environment_sm_management import SoftwareManagement
+from retry import retry
+
+
```
Comment (PradeepKiruvale) [inline tests/PySys/software_management_end_to_end/sm_apt_install_download_path/run.py:7]: Add some steps about what exactly the test does.

**Review by PradeepKiruvale — COMMENTED** [review]
Added some comments.

File: crates/common/tedge_config/src/settings.rs [hunk 4]
```diff
@@ -262,3 +262,17 @@ impl ConfigSetting for SoftwarePluginDefaultSetting {
...
 }
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct TmpPathDefaultSetting;
+
+impl ConfigSetting for TmpPathDefaultSetting {
+    const KEY: &'static str = "tmp.path";
+
+    const DESCRIPTION: &'static str = concat!(
+        "The default download path to be used for downloads on the device",
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/settings.rs:273]:
```suggestion
        "The default temporary path to be used for downloads on the device",
```

File: crates/common/tedge_config/src/tedge_config_dto.rs [hunk 5]
```diff
@@ -88,3 +91,10 @@ pub(crate) struct MqttConfigDto {
 pub(crate) struct SoftwareConfigDto {
     pub(crate) default_plugin_type: Option<String>,
 }
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct TmpPathConfigDto {
+    #[serde(default)]
+    pub(crate) path: Option<FilePath>,
+}
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_dto.rs:100]: We add here a set of config related to `Tmp` with a single property, the `path`.
I wonder if doing the reverse would be better: A config set related to `Path` with a `tmp` property.
The latter is more likely to be extended with misc paths, while the former will be a singleton for ever.

**Review by didier-wenzek — COMMENTED** [review]
Some minor comments.

**Comment by didier-wenzek** [comment]
> I have a question regarding the naming `tmp.path`. Should it be `temp.path` instead?

I'm fine with `tmp.path`.

File: crates/core/tedge_agent/src/agent.rs [hunk 6]
```diff
@@ -380,7 +394,11 @@ impl SmAgent {
             .new_log_file(LogKind::SoftwareUpdate)
             .await?;
 
-        let response = plugins.lock().unwrap().process(&request, log_file).await; // `unwrap` should be safe here as we only access data.
+        let response = plugins
+            .lock()
+            .unwrap()
```
[the remark this answers is not in the document]
Comment (initard) [reply crates/core/tedge_agent/src/agent.rs:397]: this unwrap was already there

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**